### ISSUE TITLE
npm version not creating commit

### ISSUE THREAD

**Issue by JaKXz** [npm/npm#9111]
From https://github.com/npm/npm/issues/8620#issuecomment-126098728

I have also tried using the `--force` flag with no luck.

**Comment by JaKXz** [comment]
#9099 looks like it's related, because my directory structure is a bit weird for this project:

```
my-project-root
|----.git/
|----other-files.txt
|----subdir/
|--------package.json
|--------index.html
```

and I'm running `npm version minor -m "upgrade to %s for reasons"` from `subdir`.

**Comment by aredridel** [comment]
Oof. that layout presents a problem since npm is definitely not in charge of the whole tree, and you can't commit just a subtree in git.

**Comment by dreyescat** [comment]
Definitely the layout presents a problem. `npm version` is assuming that both `.git` directory and `package.json` are found in the same directory. As you have your `package.json` in a subdirectory, when you run `npm version` it [will fail to create a tag because it is not able to find the `.git` directory](https://github.com/npm/npm/blob/master/lib/version.js#L144-L150).

As a *dirty* workaround... you could just bypass the `.git` checking just creating a fake empty `.git` folder along with your `package.json`. And when I say a fake empty folder I mean `mkdir .git` in the same folder the `package.json` file resides.

**Comment by JaKXz** [comment]
@aredridel @dreyescat thanks for the responses - making a fake `.git` directory worked and is an acceptable workaround for now.

**Comment by andrewbranch** [comment]
Yeah, I think it would be nice to enable this through an option or something. We have a single repo that contains many npm packages (for reasons that we just have to deal with for now), and I thought to myself, "I can still make `npm version` work pretty nicely by making project-specific values for `tag-version-prefix` and `messag` in each package's .npmrc. Not so. 😢

**Comment by adambuczynski** [comment]
I will +1 on this, and I think I have a valid use case. My folder structure looks approximately like this:

```
project
|_ client
|_ server
```

The `server` folder contains a Node Express app, with it's own `package.json` in the sub directory. The `client` folder contains an Angular application, again with it's own `package.json` in the sub directory. 

Because the Angular and Node apps have vastly different requirements in terms of dependencies, dev dependencies, run scripts, etc. it's not an option to use one shared `package.json`. However, I do want to keep both the client and server apps in one git repository, because it's one project and splitting it up in two repositories will create too much unnecessary overhead.

So... with that in mind comes the desire to run `npm version` for either the server or client, but because they are in a subdirectory I am experiencing the same problem that no commit is being done.

Is it possible, either by default or via a flag, to have `npm version` simply check if there is a `.git` present in one of the parent directories?

**Comment by ivos** [comment]
Elaborating on @adambuczynski I can't just create the `.git` dir and leave it there as it breaks my IDE (namely IDEA) which then starts to think that it is the proper `.git` dir  and fails to find the files there.

I can't create it in `preversion` hook and delete it again in `postversion` hook either as it is obviously not enough to cheat node to make the tag.

So I ended up with a bash script that wraps the `npm version $1` into `mkdir` and `rmdir`. But that's not  a meh anymore, that's a sad face. It's been here for well past a year now, it would be great if it gets resolved, somehow. I concur with the proposal above to check also parent dirs.

**Comment by adambuczynski** [comment]
I actually created a bash script that I could run from the project's main directory, which would update the versions for all subfolders and then tag the commit and push it upstream. It worked alright, although it required a temporary text file. Maybe it can be improved to not require this file, but my bash knowledge is not that advanced. For reference, this is what I used:

```
#!/usr/bin/env bash

git update-index -q --refresh
if ! git diff-index --quiet HEAD --; then
  echo "Working directory not clean, please commit your changes first"
  exit
fi

bump=$1

cd server
npm --no-git-tag-version version "$bump"
cd ../shop
npm --no-git-tag-version version "$bump"
cd ../cms
npm --no-git-tag-version version "$bump" > ../version.txt
cd ..

git add .
git commit -m "$(< version.txt)"
git push

cd server
npm run deploy:staging
cd ../shop
npm run deploy:staging
cd ../cms
npm run deploy:staging
cd ..

rm version.txt
```

This particular script would process 3 sub folders with individual components of the app.

However, I have since bitten the bullet and split up all our projects that had a similar structure into separate repositories after all. While this adds some overhead, I have found it does come with some benefits and it plays a bit nicer with Atom, npm and Heroku. So we've stuck with that.
